### PR TITLE
Add Sentry debug command and coverage tests

### LINKED ISSUE #1479: Sentry coverage 5/5 — Tests, fingerprint rules, and "opensre debug sentry" doctor command
Part 5 of 5 in the Sentry coverage rollout. Locks in the new behavior with tests, makes Sentry issue grouping clean, and gives users a one-shot DSN smoke test.

## Background

PRs 1–4 widen Sentry coverage substantially. This PR adds the safety net so that future regressions are caught and Sentry's UI stays usable as event volume rises.

### Tests

- [ ] Extend [`tests/test_sentry_init.py`](https://github.com/Tracer-Cloud/opensre/blob/main/tests/test_sentry_init.py) to cover PR 1 changes:
    - integrations list contains `LoggingIntegration`, `AsyncioIntegration`, `HttpxIntegration`.
    - `in_app_include=[\"app\"]` is set.
    - `entrypoint`, `runtime`, `deployment_method` tags are set on the global scope after `init_sentry(entrypoint=\"...\")`.
    - New scrubber paths: `request.data` redaction, prompt-key redaction, http-breadcrumb header strip.
    - Remember `_init_sentry_once.cache_clear()` between cases (it's `@cache`d).
- [ ] Extend [`tests/pipeline/test_runners_sentry.py`](https://github.com/Tracer-Cloud/opensre/blob/main/tests/pipeline/test_runners_sentry.py) with a LangGraph-node-internal exception case. Lock in the PR 2(b) `_traced_node` wrapper: an exception raised inside a node should produce exactly one captured event tagged `node=<name>`.
- [ ] New `tests/tools/test_basetool_sentry.py`:
    - Define a fake `BaseTool` subclass whose `run()` raises.
    - Assert that calling the tool triggers `capture_exception` exactly once, sets the `tool` tag, and returns `{\"error\": ..., \"exception_type\": ...}`.
    - Same coverage for the `@tool` decorator path.

### Fingerprint rules

- [ ] In `_before_send` (or via `sentry_sdk.set_context`), add a fingerprint rule so the `Tool error` family groups per-tool rather than collapsing into one giant issue:

    \`\`\`python
    tool_name = (event.get(\"tags\") or {}).get(\"tool\")
    if tool_name:
        event[\"fingerprint\"] = [\"tool-error\", tool_name, \"{{ default }}\"]

    node_name = (event.get(\"tags\") or {}).get(\"node\")
    if node_name:
        event[\"fingerprint\"] = [\"node-error\", node_name, \"{{ default }}\"]
    \`\`\`

- [ ] Same treatment for `subsystem`/`integration` tags from PR 4 if needed (decide based on real Sentry grouping after a week of data).

### Doctor command

- [ ] Add `opensre debug sentry` (or `make sentry-doctor`) that:
    1. Calls `init_sentry(entrypoint=\"debug\")`.
    2. Captures a synthetic exception via `capture_exception(...)`.
    3. Calls `sentry_sdk.flush(timeout=5)`.
    4. Prints whether the event was sent (`flush()` returns `True` if all events were transmitted) and the resolved DSN host (so users can verify their `OPENSRE_SENTRY_DSN` env override worked).
- [ ] Document in `docs/` (probably a new section in [`docs/sentry.mdx`](https://github.com/Tracer-Cloud/opensre/blob/main/docs/sentry.mdx)) so users self-hosting with a custom DSN have a one-shot verification.

## Acceptance

- All new tests pass under `make test-cov`.
- `opensre debug sentry` exits 0 and prints a Sentry event ID when DSN is set; exits non-zero when DSN is unreachable / disabled.
- After a week of production data, Sentry issue list is filterable by `tool`, `node`, `integration` tags and groupings make sense.

## Risks

- Fingerprint rules can over- or under-group. Start with `tool` + `\"{{ default }}\"` so Sentry's default grouping still applies inside a tool — easy to tighten later.
- The doctor command must not capture real user telemetry; tag the synthetic exception with `debug=true` so it can be filtered out in Sentry rules.

This is part of a 5-PR Sentry coverage rollout. Best done after [PRs 1 (#1475)](https://github.com/Tracer-Cloud/opensre/issues/1475) and [2 (#1476)](https://github.com/Tracer-Cloud/opensre/issues/1476).

### PR DESCRIPTION
**Pull request by amir-h-rassafi**
Fixes #1479

#### Describe the changes you have made in this PR -
- Add Sentry grouping for OpenSRE tool and LangGraph node exceptions.
- Capture tool failures with explicit `tool` tags from both `BaseTool` and `@tool` paths.
- Add `opensre debug sentry` as a synthetic smoke check for DSN delivery.
- Extend Sentry tests around init options, redaction, fingerprints, tool failures, node failures, and the debug command.
- Add a short Sentry docs section for the debug command.

### Demo/Screenshot for feature changes and bug fixes -
Live Sentry smoke demo:

image

image

image

Local verification:

```text
make lint
make format-check
make typecheck
make test-cov
```

`make test-cov` result:

```text
6809 passed, 5 skipped, 118 warnings in 94.15s
```

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**
Draft for self-review before requesting review.

The change keeps Sentry behavior centralized in `app/utils/sentry_sdk.py`: event tagging, scrubbing, and fingerprinting are applied before events leave the process. Tool wrappers pass explicit tags when exceptions are captured, and pipeline node execution uses a traced wrapper so node-internal exceptions are captured once with the node name.

The debug command initializes Sentry with the `debug` entrypoint, sends a synthetic exception tagged `debug=true`, flushes with a short timeout, and returns a non-zero exit when telemetry is disabled, Sentry does not return an event ID, or a transport explicitly reports that pending work was not flushed.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by amir-h-rassafi** [comment]
@greptile review

**Review by VibhorGautam — COMMENTED** [review]
resolved_sentry_dsn_host is the right boundary for the debug command, exposes only the host not the credential-bearing DSN

one detail in _apply_fingerprint_rules: when an event has both tool and node tags, the node fingerprint wins because the checks are sequential. that precedence isnt documented or tested, and tool can be the more actionable grouping key for failures inside a broad node. id either make the precedence explicit with a comment and test, or preserve both signals in the fingerprint

flush-based verification in the debug command test is useful coverage

File: app/cli/commands/debug.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        init_sentry(entrypoint="debug")
+    except Exception as exc:
+        click.echo(f"Sentry init failed: {type(exc).__name__}: {exc}", err=True)
+        raise SystemExit(1) from exc
+
+    try:
+        import sentry_sdk
+    except ModuleNotFoundError as exc:
+        click.echo("Sentry SDK is not installed.", err=True)
+        raise SystemExit(1) from exc
```
[the remark this answers is not in the document]
Comment (amir-h-rassafi) [reply app/cli/commands/debug.py]: Fixed in 40e8a756.

File: app/utils/sentry_sdk.py [hunk 2]
```diff
@@ -473,23 +501,27 @@ def capture_exception(
     *,
     context: str | None = None,
     extra: Mapping[str, Any] | None = None,
-) -> None:
+    tags: Mapping[str, str] | None = None,
+) -> Any | None:
```
[the remark this answers is not in the document]
Comment (amir-h-rassafi) [reply app/utils/sentry_sdk.py]: Fixed in c4c5ce77.

File: app/cli/__main__.py [hunk 3]
```diff
@@ -208,6 +208,13 @@ def _is_update_invocation(argv: list[str]) -> bool:
     return bool(command_parts) and command_parts[0] == "update"
 
 
+def _sentry_entrypoint_for_invocation(argv: list[str]) -> str:
+    command_parts = _resolve_command_parts(cli, argv)
+    if command_parts == ["debug", "sentry"]:
+        return "debug"
+    return "cli"
```
[the remark this answers is not in the document]
Comment (amir-h-rassafi) [reply app/cli/__main__.py:215]: Fixed in 6c040974.

File: app/utils/sentry_sdk.py [hunk 4]
```diff
@@ -248,6 +261,20 @@ def _event_has_operator_actionable_llm_error(event: dict[str, Any]) -> bool:
...
+    if not isinstance(tags, dict):
+        return
+
+    tool_name = tags.get("tool")
+    if isinstance(tool_name, str) and tool_name:
+        event["fingerprint"] = ["tool-error", tool_name, "{{ default }}"]
+
+    node_name = tags.get("node")
+    if isinstance(node_name, str) and node_name:
+        event["fingerprint"] = ["node-error", node_name, "{{ default }}"]
```
[the remark this answers is not in the document]
Comment (amir-h-rassafi) [reply app/utils/sentry_sdk.py:276]: Fixed in 47359869.

**Review by VibhorGautam — COMMENTED** [review]
_traced_node only wraps resolve_integrations and extract_alert, are the rest of the pipeline stages coming in a follow-up? would be weird if half the nodes report to sentry and the other half dont. the dedup stamp on _capture_exception_once makes sense for when the same error bubbles through multiple layers

**Comment by muddlebee** [comment]
demo is missing and reviews not fixed yet, check score 4/5

**Comment by amir-h-rassafi** [comment]
@greptile review

**Comment by muddlebee** [comment]
@cursor check CI

**Comment by muddlebee** [comment]
@amir-h-rassafi nice work :+1:  LGTM
